### DISCUSSION TITLE
[Bokeh should work with Polars DataFrame]

### DISCUSSION THREAD

**Discussion by vahsi-bozkirli** [bokeh/bokeh#12727] [General]
### Problem description

ı see videos about polars python library better pandas python library for data management at nowadays and there is reviews about polars better works on data but today when ı tried visualize polars dataframe with bokeh ı realize bokeh expected pandas dataframe for visualize

```
csvFile = Path("mtcars.csv")
df = pl.read_csv(csvFile)
df.head()
fig = figure(width=400, height=400)
fig.circle(x="cyl",y="disp",source=df,
         size=20, 
         line_color="navy", 
         fill_color="navy",
         fill_alpha=0.5)
show(fig)
```
df is a polars dataframe and codes gives me this error
```
ValueError: expected a dict or pandas.DataFrame, got shape: (32, 12)
```

### Feature description

when we work on visualize data with polars python library and bokeh python library , we should not get error 

### Potential alternatives

bokeh should visualize every dataframe

**Comment by bryevdv** [comment]
[converting to a GH discussion as this is a very preliminary proposal]

> Bokeh expected pandas dataframe for visualize

@vbtuna this is not true. Fundamentally, Bokeh expects a dictionary that maps column names to sequences, and Bokeh is fairly casual about what types constitute a sequence. *Additionally*, Bokeh also explicitly recognizes and handle Pandas data frames as a *convenience*, owing to Pandas widespread popularity. It looks like `pypolars` has a `to_dict` method on its dataframes:

https://pola-rs.github.io/polars/py-polars/html/reference/dataframe/api/polars.DataFrame.to_dict.html#polars.DataFrame.to_dict

I think there is a fairly decent chance that if you pass the return value of `to_dict()` to a Bokeh `ColumnDataSource`, that it might work. Please give it a try and report back. Even if it does not work out of the box, there may be simple changes to allow the `to_dict` output to work with Bokeh via duck-typing. 

It's unlikely that we would spend effort to add explicit support for `pypoloars` (i.e. ever import `pypolars` inside Bokeh) until there is a much stronger demand signal from users or evidence of it becoming very widespread, on the order of pandas, numpy etc.

**Reply by vahsi-bozkirli** [reply]
thanks for solution, when ı write this codes this works for me.
```
fig = figure(width=400, height=400)
fig.circle(x="cyl",y="disp",source=df.to_dict(as_series=False),
         size=20, 
         line_color="navy", 
         fill_color="navy",
         fill_alpha=0.5)
show(fig)
```
**df.to_dict(as_series=False)** output is:
```
{'': ['Mazda RX4',
  'Mazda RX4 Wag',
  'Datsun 710',
  'Hornet 4 Drive',
  'Hornet Sportabout',
  'Valiant',
  'Duster 360',
  'Merc 240D',
  'Merc 230',
  'Merc 280',
  ....
'mpg': [21.0,
  21.0,
  22.8,
  21.4,
  18.7,
  18.1,
  14.3,
  24.4,
  22.8,
  19.2,
  17.8,
and same pattern for another columns....
```

**Comment by hoxbro** [comment]
Support for polars was added in version 3.8.
